src/source/shift.rs

```diff
@@ -4,141 +4,136 @@ use json;
+        let child = node_config.create_cmd(name)
             .stdin(Stdio::piped())
             .stdout(Stdio::piped())
-            .stderr(Stdio::piped())
+            .stderr(Stdio::inherit())
             .spawn()
             .map_err(Error::CouldNotLaunch)?;
 
-        if let Some(ref mut stdin) = child.stdin {
-            stdin.write(script.as_bytes())
-                .map_err(Error::ExecutionError)?;
-        }
-
-        let output = child.wait_with_output()
-            .map_err(Error::ExecutionError)?;
+        Ok(Script(Mutex::new(ScriptIO {
+            input: LineWriter::new(child.stdin.unwrap()),
+            output: BufReader::new(child.stdout.unwrap()).lines(),
+        })))
+    }
 
-        debug!(target: "Shift", "Output {:?}", output);
-        if !output.status.success() {
-            error!(target: "Shift", "Script error {}", String::from_utf8(output.stderr).unwrap());
-            return Err(Error::ReturnedError(output.status));
+    pub fn transform(&self, input: &JSON) -> Result<JSON, Error> {
+        let output = (move || {
+            let mut io = self.0.lock().unwrap();
+            input.write(&mut io.input)?;
+            writeln!(io.input)?;
+            io.output.next().unwrap()
+        })().map_err(Error::IOError)?;
+
+        let result = json::parse(&output).map_err(Error::JsonError)?;
+        if let JSON::Object(mut obj) = result {
+            if let (Some(mut value), Some(ty)) = (obj.remove("value"), obj.get("type").and_then(|ty| ty.as_str())) {
+                match ty {
+                    "Ok" => return Ok(value),
+                    "Err" => if let Some(msg) = value.take_string() {
+                        return Err(Error::ParsingError(msg));
+                    },
+                    _ => {},
+                }
+            }
         }
-
-        let result = String::from_utf8(output.stdout)
-            .map_err(Error::InvalidUTF8)?;
-        Ok(result)
+        Err(Error::JsonError(json::Error::wrong_type("Result-like JSON object")))
     }
+}
 
-    fn parse_script_json_output(&self, script: &str) -> Result<JSON, Error> {
-        let stdout = self.parse_script_output(script)?;
+/// Using a Node + Shift binary to parse an AST.
+pub struct Shift {
+    parse_str: Script,
+    parse_file: Script,
+    codegen: Script,
+}
 
-        // Now attempt to parse JSON
-        json::parse(&stdout)
-            .map_err(Error::JsonError)
+impl Shift {
+    pub fn try_new() -> Result<Self, Error> {
+        let node = NodeConfig::try_new()?;
+
+        Ok(Self {
+            parse_str: Script::try_new(&node, "parse_str.js")?,
+            parse_file: Script::try_new(&node, "parse_file.js")?,
+            codegen: Script::try_new(&node, "codegen.js")?,
+        })
     }
 
-    pub fn to_source(&self, syntax: &Spec, ast: &JSON) -> Result<String, Error> {
-        let mut ast = ast.clone();
-
+    pub fn to_source(&self, syntax: &Spec, mut ast: JSON) -> Result<String, Error> {
```

Comment:
FWIW `mut` is not really part of API - it's needed internally so that Rust would know we can mutate that JSON, but for consumer and in docs it's still the same take-by-value ownership semantics.

Reply:
Sure. It's the ownership transfer that needs to be documented.

Reply:
Ah ok.